src/gsMultiGrid/gsGridHierarchy.h

```diff
@@ -0,0 +1,373 @@
+        );
+    }
+
+
+    /// @brief This function sets up a grid hierarchy by coarsening
+    ///
+    /// @param mBasis                    The gsMultiBasis to be coarsened (initial basis)
+    /// @param boundaryConditions        The boundary conditions
+    /// @param assemblerOptions          A gsOptionList defining a "DirichletStrategy" and a "InterfaceStrategy"
+    /// @param levels                    The maximum number of levels
+    /// @param degreesOfFreedom          Number of dofs in the coarsest grid in the grid hierarchy
+    ///
+    /// The algorithm terminates if either the number of levels is reached or the number of degrees of freedom
+    /// is below the given threshold.
+    ///
+    static gsGridHierarchy buildByCoarsening(
+        gsMultiBasis<T> mBasis,
+        const gsBoundaryConditions<T>& boundaryConditions,
+        const gsOptionList& assemblerOptions,
+        index_t levels,
+        index_t degreesOfFreedom = 0
+        );
+
+    /// @brief This function sets up a grid hierarchy by coarsening
+    ///
+    /// @param mBasis                    The gsMultiBasis to be coarsened (initial basis)
+    /// @param boundaryConditions        The boundary conditions
+    /// @param options                   A gsOptionList defining the necessary infomation
+    ///
+    static gsGridHierarchy buildByCoarsening(
+        gsMultiBasis<T> mBasis,
+        const gsBoundaryConditions<T>& boundaryConditions,
+        const gsOptionList& options
+    )
+    {
+        return gsGridHierarchy::buildByCoarsening(
+            give(mBasis),
+            boundaryConditions,
+            options,
+            options.askInt( "Levels", 3 ),
+            options.askInt( "DegreesOfFreedom", 0 )
+        );
+    }
+
+    /// Get the default options
+    static gsOptionList defaultOptions()
+    {
+        gsOptionList opt;
+        opt.addInt( "DirichletStrategy", "Method for enforcement of Dirichlet BCs [11..14]", 11 );
+        opt.addInt( "InterfaceStrategy", "Method of treatment of patch interfaces [0..3]", 1  );
+        opt.addInt( "Levels", "Number of levels to be constructed in the grid hierarchy", 3 );
+        opt.addInt( "DegreesOfFreedom",   "Number of dofs in the coarsest grid in the grid hierarchy (only buildByCoarsening)", 0 );
+        opt.addInt( "NumberOfKnotsToBeInserted", "The number of knots to be inserted (only buildByRefinement)", 1 );
+        opt.addInt( "MultiplicityOfKnotsToBeInserted",   "The multiplicity of the knots to be inserted (only buildByRefinement)", 1 );
+        return opt;
+    }
+
+    /// Reset the object (to save memory)
+    void clear()
+    {
+        //m_boundaryConditions.clear();
+        //m_assemblerOptions.clear();
+        m_mBases.clear();
+        m_transferMatrices.clear();
+        m_localTransferMatrices.clear();
+    }
+
+    /// Get the vector of multi bases (by reference)
+    std::vector< gsMultiBasis<T> >& getMultiBases() { return m_mBases; }
```

Comment:
Might be safer if non const references are not provided (unless they are used already), same for the other getters..